### PR TITLE
feat(schemas): make schemas mutables [EX-5435]

### PR DESCRIPTION
**Pull request by diegopf**
## Motivation and context
Create a new type MutableSchema and a function makeSchemaMutable that allows to create a new MutableSchema from a Schema. MutableSchemas have two additionally methods that allow us to either replace or override the original schema.

- [ ] Dependencies. If any, specify:
- [x] Open issue. If applicable, link:  [EX-5435](https://searchbroker.atlassian.net/browse/EX-5435)

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that causes existing functionality to not work as expected)
- [ ] Change requires a documentation update

## What is the destination branch of this PR?

- [x] `Main`
- [ ] Other. Specify:

## How has this been tested?

Tests performed according to [testing guidelines](./contributing/tests.md): 

## Checklist:

- [x] My code follows the **[style guidelines](./CONTRIBUTING.md#style-guides)** of this project.
- [x] I have performed a **self-review** on my own code.
- [x] I have **commented** my code, particularly in hard-to-understand areas.
- [ ] I have made corresponding changes to the **[documentation](./CONTRIBUTING.md#documentation-style-guide)**.
- [x] My changes generate **no new warnings**.
- [x] I have added **tests** that prove my fix is effective or that my feature works.
- [x] New and existing **unit tests pass locally** with my changes.

### CODE REVIEW

File: packages/x-adapter/src/mappers/schema-mapper.factory.ts [hunk 1]
```diff
@@ -15,7 +16,7 @@ import { extractValue } from '../utils/extract-value';
  * @public
  */
 export function schemaMapperFactory<Source, Target>(
-  schema: Schema<Source, Target>
+  schema: Schema<Source, Target> | MutableSchema<Schema<Source, Target>>
```
[the remark this answers is not in the document]
Comment (diegopf) [reply packages/x-adapter/src/mappers/schema-mapper.factory.ts]: Good point, you are right! At some point during development was needed but seems it is not anymore.

File: packages/x-adapter/src/schemas/schemas.types.ts [hunk 2]
```diff
@@ -48,6 +48,47 @@ export type Schema<Source = any, Target = any> = {
...
+   * Merges the original {@link Schema | schema} with the given one.
+   *
+   * @param newSchema - The {@link Schema | schema} to use to merge with the original one.
+   * @returns The {@link Schema | schema} returned by the merge.
+   */
+  $override: <Source, Target>(
+    newSchema: Schema<Source, Target>
+  ) => MutableSchema<Schema<Source, Target>>;
+  /**
+   * Allows to create a new {@link Schema | schema} using the original one as starting point.
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/schemas/schemas.types.ts:80]:
```suggestion
   * Create a new {@link Schema | schema} using the original one as starting point.
```

File: packages/x-adapter/src/schemas/schemas.types.ts [hunk 3]
```diff
@@ -48,6 +48,47 @@ export type Schema<Source = any, Target = any> = {
...
 
+/**
+ * A {@link Schema | schema} with extended functionality. In addition to
+ * regular {@link Schema | schema} includes some methods to be able to completely replace
+ * the original schema, to partially override it or create a new one.
+ *
+ * @param OriginalSchema - The {@link Schema | schema} that will be mutable.
+ *
+ *
+ */
```
Comment (LuisMartinez15) [inline packages/x-adapter/src/schemas/schemas.types.ts:59]:
```suggestion
/**
 * A {@link Schema | schema} with extended functionality to: completely replace
 * the original schema, partially override it or create a new one.
 *
 * @param OriginalSchema - The {@link Schema | schema} that will be mutable.
 *
 * @public
 */
```

**Review by LuisMartinez15 — CHANGES_REQUESTED** [review]
good job!

File: packages/x-adapter/src/schemas/utils.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+  return {
+    ...schema,
+    $replace: function <Source = any, Target = any>(
+      newSchema: Schema<Source, Target>
+    ): MutableSchema<Schema<Source, Target>> {
+      Object.keys(this).forEach(key => {
+        if (isInternalMethod(key)) {
+          return;
+        }
+        delete this[key];
```
[the remark this answers is not in the document]
Comment (diegopf) [reply packages/x-adapter/src/schemas/utils.ts:28]: Funny, I discussed the same thing with @tajespasarela. IMO I prefer the early return approach and avoid having the conditional clause negative.

File: packages/x-adapter/src/schemas/utils.ts [hunk 5]
```diff
@@ -0,0 +1,56 @@
+import { deepMerge } from '@empathyco/x-deep-merge';
+import { MutableSchema, Schema } from './schemas.types';
+
+/**
+ * Collection of internal method names for {@link MutableSchema | mutable schemas}.
+ */
+export const mutableSchemasInternalMethods = ['$replace', '$override', '$extends'];
```
[the remark this answers is not in the document]
Comment (diegopf) [reply packages/x-adapter/src/schemas/utils.ts]: You are right! Previously I was using it on another place but now does not make sense to export it. However I'd like to keep the const IMO makes the code more readable.

File: packages/x-adapter/src/schemas/schemas.types.ts (hunk 2 above)
Comment (diegopf) [reply packages/x-adapter/src/schemas/schemas.types.ts]: Sounds good to me

File: packages/x-adapter/src/schemas/utils.ts [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+/**
+ * Creates a {@link MutableSchema | mutable schema } version of a given {@link Schema | schema}.
+ *
+ * @param schema - The {@link Schema | schema} to make mutable.
+ *
+ * @returns A {@link MutableSchema | mutable schema} version of the given {@link Schema | schema}.
+ *
+ * @public
+ */
+export function makeSchemaMutable<T extends Schema>(schema: T): MutableSchema<T> {
```
[the remark this answers is not in the document]
Comment (diegopf) [reply packages/x-adapter/src/schemas/utils.ts]: Sounds good to me

File: packages/x-adapter/src/schemas/utils.ts [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+      newSchema: Schema<Source, Target>
+    ): MutableSchema<Schema<Source, Target>> {
+      return deepMerge(this, newSchema);
+    },
+    $extends: function <Source = any, Target = any>(
+      newSchema: Schema<Source, Target>
+    ): MutableSchema<Schema<Source, Target>> {
+      return deepMerge({}, this, newSchema);
+    },
+    $toString: function (includeInternalMethods = false) {
```
Comment (tajespasarela) [inline packages/x-adapter/src/schemas/utils.ts:43]: rename to standard `toString` to coercing directly to a string:

`console.log(schema)`

**Review by tajespasarela — CHANGES_REQUESTED** [review]

File: packages/x-adapter/src/schemas/utils.ts [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    ): MutableSchema<Schema<Source, Target>> {
+      return deepMerge(this, newSchema);
+    },
+    $extends: function <Source = any, Target = any>(
+      newSchema: Schema<Source, Target>
+    ): MutableSchema<Schema<Source, Target>> {
+      return deepMerge({}, this, newSchema);
+    },
+    toString: function (includeInternalMethods = false) {
+      return serialize(this, !!includeInternalMethods);
```
Comment (tajespasarela) [inline packages/x-adapter/src/schemas/utils.ts:45]: you can remove the `!!`

**Review by tajespasarela — CHANGES_REQUESTED** [review]

**Review by tajespasarela — APPROVED** [review]

**Review by LuisMartinez15 — APPROVED** [review]
good job! 👏🏿

**Pull request merged**